src/main/java/net/minecraftforge/server/console/TerminalHandler.java

```diff
@@ -19,63 +19,72 @@
 
 package net.minecraftforge.server.console;
 
-import java.io.IOException;
+import net.minecraftforge.server.terminalconsole.TerminalConsoleAppender;
 
-import org.apache.logging.log4j.LogManager;
-import org.apache.logging.log4j.Logger;
-
-import jline.console.ConsoleReader;
 import net.minecraft.server.dedicated.DedicatedServer;
-import net.minecraft.util.text.TextFormatting;
-import net.minecraftforge.server.console.log4j.TerminalConsoleAppender;
+import org.jline.reader.EndOfFileException;
+import org.jline.reader.LineReader;
+import org.jline.reader.LineReaderBuilder;
+import org.jline.reader.UserInterruptException;
+import org.jline.terminal.Terminal;
 
 public final class TerminalHandler
 {
 
-    private static final Logger logger = LogManager.getLogger();
-
     private TerminalHandler()
     {
     }
 
     public static boolean handleCommands(DedicatedServer server)
     {
-        final ConsoleReader reader = TerminalConsoleAppender.getReader();
-        if (reader != null)
-        {
-            TerminalConsoleAppender.setFormatter(new ConsoleFormatter());
-            reader.addCompleter(new ConsoleCommandCompleter(server));
+        final Terminal terminal = TerminalConsoleAppender.getTerminal();
+        if (terminal == null)
+            return false;
+
+        LineReader reader = LineReaderBuilder.builder()
+                .appName("Forge")
+                .terminal(terminal)
+                .completer(new ConsoleCommandCompleter(server))
+                .build();
+        reader.unsetOpt(LineReader.Option.INSERT_TAB);
```

Comment:
Need to call `reader.setOpt(LineReader.Option.DISABLE_EVENT_EXPANSION);` additionally here to avoid https://github.com/PaperMC/Paper/issues/1171 for Forge. (This is equivalent to `reader.setExpandEvents(false)` in the old JLine 2.x code)

Reply:
Thanks, done